src/compiler/transformers/classFields.ts

```diff
@@ -506,6 +786,165 @@ namespace ts {
+                createAssignment(
+                    weakMapName,
+                    createNew(
+                        createIdentifier("WeakMap"),
+                        /*typeArguments*/ undefined,
+                        []
+                    )
+                )
+            );
+        }
+
+        function accessPrivateIdentifier(name: PrivateIdentifier) {
+            for (let i = privateIdentifierEnvironmentStack.length - 1; i >= 0; --i) {
+                const env = privateIdentifierEnvironmentStack[i];
+                if (env.has(name.escapedText)) {
+                    return env.get(name.escapedText);
+                }
+            }
+            return undefined;
+        }
+
+
+        function wrapPrivateIdentifierForDestructuringTarget(node: PrivateIdentifierPropertyAccessExpression) {
+            return createPropertyAccess(
+                createObjectLiteral([
+                    createSetAccessor(
+                        /*decorators*/ undefined,
+                        /*modifiers*/ undefined,
+                        "value",
+                        [createParameter(
+                            /*decorators*/ undefined,
+                            /*modifiers*/ undefined,
+                            /*dotDotDotToken*/ undefined, "x",
+                            /*questionToken*/ undefined,
+                            /*type*/ undefined,
+                            /*initializer*/ undefined
+                        )],
+                        createBlock(
+                            [createExpressionStatement(
+                                visitNode(
+                                    createAssignment(node, createIdentifier("x")),
+                                    visitor
+                                )
+                            )]
+                        )
+                    )
+                ]),
+                "value"
+            );
+        }
+
+        function transformDestructuringAssignmentTarget(node: ArrayLiteralExpression | ObjectLiteralExpression) {
+            const hasPrivateIdentifiers = isArrayLiteralExpression(node) ?
+                forEach(node.elements, isPrivateIdentifierPropertyAccessExpression) :
+                forEach(node.properties, property => isPropertyAssignment(property) && isPrivateIdentifierPropertyAccessExpression(property.initializer));
+            if (!hasPrivateIdentifiers) {
+                return node;
+            }
+            if (isArrayLiteralExpression(node)) {
+                // Transforms private names in destructuring assignment array bindings.
+                //
+                // Source:
+                // ([ this.#myProp ] = [ "hello" ]);
+                //
+                // Transformation:
+                // [ { set value(x) { this.#myProp = x; } }.value ] = [ "hello" ];
+                return updateArrayLiteral(
+                    node,
+                    node.elements.map(

```

Comment:
```suggestion
                    visitNodes(node.elements, visitArrayAssignmentTarget, isExpression)
```

Where `visitArrayAssignmentTarget` is essentially the arrow function below, except you should call `visitor` on `expr` when its not a `PrivateIdentifierPropertyAccessExpression`.

Then you won't need the guard at the top of the function.